File: virtual_keyboard.rb
#!/usr/bin/env ruby
# virtual_keyboard.rb
# encoding: UTF-8

require 'io/console'

COLORS = {
  reset: "\e[0m",
  green: "\e[92m",
  blue: "\e[94m",
  yellow: "\e[93m",
  bold: "\e[1m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

KEYBOARD_EN = [
  %w[1 2 3 4 5 6 7 8 9 0],
  %w[Q W E R T Y U I O P],
  %w[A S D F G H J K L],
  %w[Z X C V B N M]
]

KEYBOARD_RU = [
  %w[1 2 3 4 5 6 7 8 9 0],
  %w[Й Ц У К Е Н Г Ш Щ З],
  %w[Ф Ы В А П Р О Л Д],
  %w[Я Ч С М И Т Ь]
]

class VirtualKeyboard
  attr_accessor :lang, :text, :shift, :layout

  def initialize(lang = 'en')
    @lang = lang
    @text = []
    @shift = false
    @layout = lang == 'en' ? KEYBOARD_EN : KEYBOARD_RU
  end

  def display
    system('clear') || system('cls')
    puts colorize('🖥️  ВИРТУАЛЬНАЯ КЛАВИАТУРА', :bold)
    puts colorize('Управление: номер клавиши → добавить, 0 → пробел, - → удалить, . → выход', :yellow)
    puts colorize('8 → переключить раскладку, 9 → переключить регистр (Shift)\n', :yellow)
    puts colorize("Текст: #{@text.join}", :green)
    puts

    idx = 1
    @layout.each do |row|
      line = ''
      row.each do |ch|
        display_ch = @shift ? ch.upcase : ch.downcase
        line += colorize(idx.to_s.rjust(2, '0'), :blue) + ':' + colorize(display_ch, :blue) + '  '
        idx += 1
      end
      puts line
    end
    puts "\nСпециальные: 0-пробел, --удалить, .-выход, 8-язык, 9-Shift"
  end

  def run
    loop do
      display
      print colorize('Введите номер команды: ', :bold)
      cmd = gets.chomp.strip
      case cmd
      when '.'
        break
      when '-'
        @text.pop if @text.any?
      when '0'
        @text << ' '
      when '8'
        @lang = @lang == 'en' ? 'ru' : 'en'
        @layout = @lang == 'en' ? KEYBOARD_EN : KEYBOARD_RU
      when '9'
        @shift = !@shift
      else
        num = cmd.to_i
        if num > 0
          flat = @layout.flatten
          if num <= flat.size
            ch = flat[num-1]
            ch = @shift ? ch.upcase : ch.downcase
            @text << ch
          else
            puts colorize('Неверный номер!', :red)
          end
        else
          puts colorize('Неверный ввод!', :red)
        end
      end
    end
    puts colorize("\nИтоговый текст: #{@text.join}", :green)
  end
end

if __FILE__ == $0
  kb = VirtualKeyboard.new('en')
  kb.run
end

File: virtual_keyboard.py
# virtual_keyboard.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import sys
import time

# ANSI-цвета
COLORS = {
    'reset': '\033[0m',
    'green': '\033[92m',
    'blue': '\033[94m',
    'yellow': '\033[93m',
    'bold': '\033[1m'
}

def colorize(text, color):
    return f"{COLORS.get(color, '')}{text}{COLORS['reset']}"

# Клавиатурные раскладки (три ряда)
KEYBOARD_EN = [
    ['1', '2', '3', '4', '5', '6', '7', '8', '9', '0'],
    ['Q', 'W', 'E', 'R', 'T', 'Y', 'U', 'I', 'O', 'P'],
    ['A', 'S', 'D', 'F', 'G', 'H', 'J', 'K', 'L'],
    ['Z', 'X', 'C', 'V', 'B', 'N', 'M']
]

KEYBOARD_RU = [
    ['1', '2', '3', '4', '5', '6', '7', '8', '9', '0'],
    ['Й', 'Ц', 'У', 'К', 'Е', 'Н', 'Г', 'Ш', 'Щ', 'З'],
    ['Ф', 'Ы', 'В', 'А', 'П', 'Р', 'О', 'Л', 'Д'],
    ['Я', 'Ч', 'С', 'М', 'И', 'Т', 'Ь']
]

class VirtualKeyboard:
    def __init__(self, lang='en'):
        self.lang = lang
        self.text = []
        self.shift = False
        self.layout = KEYBOARD_EN if lang == 'en' else KEYBOARD_RU

    def display(self):
        os.system('clear' if os.name == 'posix' else 'cls')
        print(colorize("🖥️  ВИРТУАЛЬНАЯ КЛАВИАТУРА", 'bold'))
        print(colorize("Управление: номер клавиши → добавить, 0 → пробел, - → удалить, . → выход", 'yellow'))
        print(colorize("8 → переключить раскладку, 9 → переключить регистр (Shift)\n", 'yellow'))
        print(colorize("Текст: " + ''.join(self.text), 'green'))
        print()

        # Отображение клавиатуры с номерами
        idx = 1
        for row in self.layout:
            line = ''
            for ch in row:
                display_ch = ch.upper() if self.shift else ch.lower()
                line += f"{colorize(str(idx).zfill(2), 'blue')}:{colorize(display_ch, 'blue')}  "
                idx += 1
            print(line)
        print("\nСпециальные: 0-пробел, --удалить, .-выход, 8-язык, 9-Shift")

    def run(self):
        while True:
            self.display()
            cmd = input(colorize("Введите номер команды: ", 'bold')).strip()
            if cmd == '.':
                break
            if cmd == '-':
                if self.text:
                    self.text.pop()
                continue
            if cmd == '0':
                self.text.append(' ')
                continue
            if cmd == '8':
                self.lang = 'ru' if self.lang == 'en' else 'en'
                self.layout = KEYBOARD_RU if self.lang == 'ru' else KEYBOARD_EN
                continue
            if cmd == '9':
                self.shift = not self.shift
                continue
            try:
                num = int(cmd)
                # Найти символ по номеру
                flat = [ch for row in self.layout for ch in row]
                if 1 <= num <= len(flat):
                    ch = flat[num-1]
                    if self.shift:
                        ch = ch.upper()
                    else:
                        ch = ch.lower()
                    self.text.append(ch)
                else:
                    print(colorize("Неверный номер!", 'red'))
                    time.sleep(0.5)
            except ValueError:
                print(colorize("Неверный ввод!", 'red'))
                time.sleep(0.5)

        print(colorize(f"\nИтоговый текст: {''.join(self.text)}", 'green'))

if __name__ == '__main__':
    try:
        kb = VirtualKeyboard('en')
        kb.run()
    except KeyboardInterrupt:
        print(colorize("\nВыход.", 'yellow'))
        sys.exit(0)
